src/main/java/com/aws/greengrass/shadowmanager/ShadowManager.java

```diff
@@ -406,6 +410,7 @@ protected void shutdown() throws InterruptedException {
             stopSyncingShadows(true);
             database.close();
             inboundRateLimiter.clear();
+            pubSubIntegrator.unsubscribe();
```

Comment:
looking at this order of shutdown, can you switch it to:
```
stopSyncingShadows(true);
pubSubIntegrator.unsubscribe();
inboundRateLimiter.clear();
database.close();
```

It seems like we should first stop things that generate work for us (syncing, handling IPC subscriptions), then clear the rate limiter, and finally stop the database.

We don't want to close the database while we are still able to receive messages.

Also, if we are in the middle of processing an IPC message and we get shutdown, do we do anything to "finish" the IPC request?

Reply:
Right now, we don't do anything. It will just stop. Do we want to add an `AtomicInteger` to keep track of how many requests are being processed right now and only shut down if and when the count becomes 0?

Reply:
I think that ideally we could finish in flight requests. For syncing it will tell it to cancel the thread, but if a request is executing it will wait for that request to finish.

Let's not add extra complexity to this PR though